### PR TITLE
Add support for self-hosted ClickHouse

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
```yaml
clickhouse:
  enabled: true

  # Optional configuration to pull from a private registry
  image:
    registry: docker.io
    repository: bitnami/clickhouse
    tag: 24.4.1-debian-12-r4

```

> [!IMPORTANT]
> Make sure that `global.dvcx.clickHouse.dsn` is **not set**, or it will override the self-hosted ClickHouse connection string.

### CODE REVIEW

**Review by shcheklein — COMMENTED** [review]
Do we need to bring the whole chart like we do for other tools? to make it self-contained?

**Comment by 0x2b3bfa0** [comment]
> _Do we need to bring the whole chart like we do for other tools? to make it self-contained?_

What do you mean exactly by "bring[ing] the whole chart"? If you mean "vendoring" (committing) any `tgz` with external Bitnami charts to our repository, we've never done that for any of the tools.

**Comment by shcheklein** [comment]
I primarily mean the tgz bundle that we provide, not necessarily vendoring code itself. I'm not sure how exactly that is done.

**Comment by 0x2b3bfa0** [comment]
We don't have to bundle (vendor) third-party charts with ours; Helm fetches them automatically.

**Review by shcheklein — APPROVED** [review]
Linter is failing though. THanks @0x2b3bfa0 !

**Comment by shcheklein** [comment]
Q: does it come with some basic PV / PVC?

File: .github/workflows/helm-lint-and-install.yaml [hunk 1]
```diff
@@ -59,7 +59,7 @@ jobs:
         if: steps.list-changed.outputs.changed == 'true'
         run: |
           ct lint-and-install --target-branch ${{ github.event.repository.default_branch }} --upgrade --skip-missing-values --debug \
-          --helm-extra-set-args '--set ci=true --set global.blobvault.persistentVolume.storageClassName="standard" --set imagePullSecrets[0].name=iterativeai --set dockerUsername=${{ vars.ITERATIVE_DOCKER_REGISTRY_USER }} --set dockerPassword=${{ secrets.ITERATIVE_DOCKER_REGISTRY_PASSWORD }} --set dockerServer=docker.iterative.ai'
+          --helm-extra-set-args '--set ci=true --set global.blobvault.persistentVolume.storageClassName="standard" --set clickhouse.auth.password="clickhouse" --set imagePullSecrets[0].name=iterativeai --set dockerUsername=${{ vars.ITERATIVE_DOCKER_REGISTRY_USER }} --set dockerPassword=${{ secrets.ITERATIVE_DOCKER_REGISTRY_PASSWORD }} --set dockerServer=docker.iterative.ai'
```
Comment (0x2b3bfa0) [inline .github/workflows/helm-lint-and-install.yaml:62]: https://github.com/iterative/helm-charts/actions/runs/9332063406/job/25690736127#step:10:117

```
Error: UPGRADE FAILED: execution error at (studio/charts/clickhouse/templates/secret.yaml:19:21): 
PASSWORDS ERROR: You must provide your current passwords when upgrading the release.
                 Note that even after reinstallation, old credentials may be needed as they may be kept in persistent volume claims.
                 Further information can be obtained at https://docs.bitnami.com/general/how-to/troubleshoot-helm-chart-issues/#credential-errors-while-upgrading-chart-releases

    'auth.password' must not be empty, please add '--set auth.***' to the command. To get the current value:

        export ADMIN_PASSWORD=$(kubectl get secret --namespace "studio-woa6hxx2un" studio-woa6hxx2un-clickhouse -o jsonpath="{.data.admin-password}" | base64 -d)
```

**Comment by 0x2b3bfa0** [comment]
> _Q: does it come with some basic PV / PVC?_

Yes, you can configure it with [these options](https://github.com/bitnami/charts/tree/main/bitnami/clickhouse#persistence-parameters) if needed.

File: .github/workflows/helm-lint-and-install.yaml (hunk 1 above)
Comment (0x2b3bfa0) [reply .github/workflows/helm-lint-and-install.yaml:62]: @iterative/platform, is there any better solution?

**Comment by shcheklein** [comment]
@0x2b3bfa0 are there any blockers? are we good to go here?

**Comment by 0x2b3bfa0** [comment]
Everything seems to work, **but** I'm getting obscure ClickHouse errors when running queries, not sure why:

```
Using cached virtualenv
Traceback (most recent call last):
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/dbapi/cursor.py", line 111, in execute
    response = execute(
               ^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/client.py", line 382, in execute
    rv = self.process_ordinary_query(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/client.py", line 580, in process_ordinary_query
    return self.receive_result(with_column_types=with_column_types,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/client.py", line 213, in receive_result
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/result.py", line 50, in get_result
    for packet in self.packet_generator:
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/client.py", line 229, in packet_generator
    packet = self.receive_packet()
             ^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/client.py", line 246, in receive_packet
    raise packet.exception
clickhouse_driver.errors.ServerException: Code: 60.
DB::Exception: Unknown table expression identifier 'src_s3://example/_1_1' in scope SELECT `src_s3://example/_1_1`.id, `src_s3://example/_1_1`.vtype, `src_s3://example/_1_1`.dir_type, `src_s3://example/_1_1`.parent, `src_s3://example/_1_1`.name, `src_s3://example/_1_1`.etag, `src_s3://example/_1_1`.version, `src_s3://example/_1_1`.is_latest, `src_s3://example/_1_1`.last_modified, `src_s3://example/_1_1`.size, `src_s3://example/_1_1`.owner_name, `src_s3://example/_1_1`.owner_id, `src_s3://example/_1_1`.random, `src_s3://example/_1_1`.location, `src_s3://example/_1_1`.source FROM `src_s3://example/_1_1` ORDER BY `src_s3://example/_1_1`.id ASC LIMIT 20. Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c9a449b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000780b9ac
2. DB::Exception::Exception<String const&, String>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String>::type>, String const&, String&&) @ 0x00000000080b544b
3. DB::(anonymous namespace)::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x0000000010a7d9b9
4. DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x0000000010a780c5
5. DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x0000000010a76983
6. DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) (.llvm.9862110563685019565) @ 0x0000000010d0aafd
7. DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&) @ 0x0000000010d09899
8. std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> std::__function::__policy_invoker<std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>::__call_impl<std::__function::__default_alloc_func<DB::registerInterpreterSelectQueryAnalyzer(DB::InterpreterFactory&)::$_0, std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>>(std::__function::__policy_storage const*, DB::InterpreterFactory::Arguments const&) (.llvm.9862110563685019565) @ 0x0000000010d0c86c
9. DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x0000000010c9ec79
10. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001111a030
11. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x00000000111169ba
12. DB::TCPHandler::runImpl() @ 0x00000000122a59c4
13. DB::TCPHandler::run() @ 0x00000000122c1fb9
14. Poco::Net::TCPServerConnection::start() @ 0x0000000014c105b2
15. Poco::Net::TCPServerDispatcher::run() @ 0x0000000014c113f9
16. Poco::PooledThread::run() @ 0x0000000014d09a61
17. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000014d07ffd
18. ? @ 0x000079b537739134
19. ? @ 0x000079b5377b8a40


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 3, in <module>
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/query/dataset.py", line 1804, in query_wrapper
    dataset_query = dataset_query.save(name)
                    ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/query/dataset.py", line 1646, in save
    query = self.apply_steps()
            ^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/query/dataset.py", line 1163, in apply_steps
    result = query.starting_step.apply()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/query/dataset.py", line 195, in apply
    self.catalog.index([self.path], **self.kwargs)
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 2172, in index
    self.enlist_sources(
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 831, in enlist_sources
    listing, file_path = self.enlist_source(
                         ^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 750, in enlist_source
    dataset = self.create_dataset(storage_dataset_name, listing=True)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 1035, in create_dataset
    return self.create_new_dataset_version(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 1085, in create_new_dataset_version
    self.update_dataset_version_with_warehouse_info(dataset, version)
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 1115, in update_dataset_version_with_warehouse_info
    dataset_rows := list(
                    ^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/catalog/catalog.py", line 1384, in ls_dataset_rows
    yield from self.warehouse.get_dataset_rows(
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/dvcx/data_storage/warehouse.py", line 325, in get_dataset_rows
    for row in self.dataset_rows_select(query):
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhousedbadapter/clickhouse.py", line 759, in dataset_rows_select
    rows = self.db.execute(select_query, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhousedbadapter/clickhouse.py", line 510, in execute
    cursor.execute(*compiled)
  File "/tmp/local/dvcx_venv/python3.12/default/lib/python3.12/site-packages/clickhouse_driver/dbapi/cursor.py", line 117, in execute
    raise OperationalError(orig)
clickhouse_driver.dbapi.errors.OperationalError: Code: 60.
DB::Exception: Unknown table expression identifier 'src_s3://example/_1_1' in scope SELECT `src_s3://example/_1_1`.id, `src_s3://example/_1_1`.vtype, `src_s3://example/_1_1`.dir_type, `src_s3://example/_1_1`.parent, `src_s3://example/_1_1`.name, `src_s3://example/_1_1`.etag, `src_s3://example/_1_1`.version, `src_s3://example/_1_1`.is_latest, `src_s3://example/_1_1`.last_modified, `src_s3://example/_1_1`.size, `src_s3://example/_1_1`.owner_name, `src_s3://example/_1_1`.owner_id, `src_s3://example/_1_1`.random, `src_s3://example/_1_1`.location, `src_s3://example/_1_1`.source FROM `src_s3://example/_1_1` ORDER BY `src_s3://example/_1_1`.id ASC LIMIT 20. Stack trace:

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c9a449b
1. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000780b9ac
2. DB::Exception::Exception<String const&, String>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String>::type>, String const&, String&&) @ 0x00000000080b544b
3. DB::(anonymous namespace)::QueryAnalyzer::resolveQuery(std::shared_ptr<DB::IQueryTreeNode> const&, DB::(anonymous namespace)::IdentifierResolveScope&) @ 0x0000000010a7d9b9
4. DB::QueryAnalysisPass::run(std::shared_ptr<DB::IQueryTreeNode>&, std::shared_ptr<DB::Context const>) @ 0x0000000010a780c5
5. DB::QueryTreePassManager::run(std::shared_ptr<DB::IQueryTreeNode>) @ 0x0000000010a76983
6. DB::(anonymous namespace)::buildQueryTreeAndRunPasses(std::shared_ptr<DB::IAST> const&, DB::SelectQueryOptions const&, std::shared_ptr<DB::Context const> const&, std::shared_ptr<DB::IStorage> const&) (.llvm.9862110563685019565) @ 0x0000000010d0aafd
7. DB::InterpreterSelectQueryAnalyzer::InterpreterSelectQueryAnalyzer(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const> const&, DB::SelectQueryOptions const&) @ 0x0000000010d09899
8. std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> std::__function::__policy_invoker<std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>::__call_impl<std::__function::__default_alloc_func<DB::registerInterpreterSelectQueryAnalyzer(DB::InterpreterFactory&)::$_0, std::unique_ptr<DB::IInterpreter, std::default_delete<DB::IInterpreter>> (DB::InterpreterFactory::Arguments const&)>>(std::__function::__policy_storage const*, DB::InterpreterFactory::Arguments const&) (.llvm.9862110563685019565) @ 0x0000000010d0c86c
9. DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x0000000010c9ec79
10. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x000000001111a030
11. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x00000000111169ba
12. DB::TCPHandler::runImpl() @ 0x00000000122a59c4
13. DB::TCPHandler::run() @ 0x00000000122c1fb9
14. Poco::Net::TCPServerConnection::start() @ 0x0000000014c105b2
15. Poco::Net::TCPServerDispatcher::run() @ 0x0000000014c113f9
16. Poco::PooledThread::run() @ 0x0000000014d09a61
17. Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000014d07ffd
18. ? @ 0x000079b537739134
19. ? @ 0x000079b5377b8a40
```

@iterative/dvcx, any help with this would be appreciated.

**Comment by mattseddon** [comment]
@0x2b3bfa0 does the `src_s3://example/_1_1` table exist in the clickhouse database? Are you querying against `src_s3://example/` and did you index the bucket beforehand or during the query?

**Comment by 0x2b3bfa0** [comment]
Sorry for the lack of context, I was running this query from scratch, with no previous indexing:

```python
from dvcx.lib.dataset import Dataset
Dataset("s3://ldb-public/remote/models/FourClassClassifier", anon=True)
```

No tables were created on ClickHouse; the `default` database is empty after running the query.

File: charts/studio/values.yaml [hunk 2]
```diff
@@ -238,6 +238,21 @@ postgresql:
...
 
+clickhouse:
+  # -- ClickHouse enabled
+  enabled: false
+  # -- ClickHouse name override
+  fullnameOverride: studio-clickhouse
+
+  # Shards / replicas configuration
+  replicaCount: 1
+  shards: 1
```
Comment (0x2b3bfa0) [inline charts/studio/values.yaml:249]: Using more shards/replicas will cause eventual consistency issues with table creation; to be fiexed whenever we have to scale.

**Pull request merged**